app/scripts/lib/channels/duplex.js

```diff
@@ -0,0 +1,199 @@
+     */
+    send: function (command, data) {
+      var self = this;
+      return p()
+        .then(function () {
+          return self._sender.send(command, data, null);
+        });
+    },
+
+    /**
+     * Send a message, expect a response.
+     *
+     * @param {String} command
+     * @param {Object} [data]
+     * @return {Promise}
+     *        Promise will resolve when the response is received.
+     */
+    request: function (command, data) {
+      var self = this;
+
+      var messageId = this.createMessageId(command, data);
+      var outstanding = {
+        deferred: p.defer(),
+        command: command,
+        data: data,
+        messageId: messageId
+      };
+
+      // save the data beforehand in case the response is synchronous.
+      self._outstandingRequests.add(messageId, outstanding);
+
+      return p()
+        .then(function () {
+          return self._sender.send(command, data, messageId);
+        })
+        .then(function () {
+          return outstanding.deferred.promise;
+        })
+        .fail(function (err) {
+          // If there was a problem sending the request is
+          // no longer outstanding.
+          self._outstandingRequests.remove(messageId);
+          throw err;
+        });
+    },
+
+    /**
+     * Create a messageId for a given command/data combination.
+     *
+     * messageId is sent to the relier who is expected to respond
+     * with the same messageId. Used to keep track of outstanding requests.
+     *
+     * @param {String} command
+     * @param {Object} [data]
+     * @return {String}
+     */
+    createMessageId: function (command, data) { //jshint ignore: line
+      return Date.now();
+    },
+
+    onMessageReceived: function (message) {
+      var self = this;
+      var parsedMessage = this.parseMessage(message);
+      var data = parsedMessage.data;
+      var messageId = parsedMessage.messageId;
+
+      // A message is not necessarily in response to a sent request.
+      // If the message is in response to a request, then it should
+      // have a messageId.
+      var outstanding = self._outstandingRequests.get(messageId);
+      if (outstanding) {
+        self._outstandingRequests.remove(messageId);
+        outstanding.deferred.resolve(data);
+      }
+
+      // Even if the message is not in response to a request, trigger an
+      // event for any listeners that are waiting for it.
+      this.trigger(parsedMessage.command, data);
+    },
+
+    /**
+     * Parse an incoming message into `messageId`, `command`, and `data`
+     *
+     * @param {Object} message
+     * @returns {Object} parsedMessage
+     *    @param {String} parsedMessage.command - message command
+     *    @param {String} [parsedMessage.messageId] - message id, if
+     *    a response.
+     *    @param {Object} parsedMessage.data - data
+     */
+    parseMessage: function (message) {
+      // by default, the message is already parsed.
+      return message;
```

Comment:
should this return an object that the jsdoc suggests? 
i.e `return { command: message.command, ... , data: message.data }` etc